### DISCUSSION TITLE
IAB Transparency and Consent Framework

### DISCUSSION THREAD

**Discussion by rectifyer** [orestbida/cookieconsent#289] [Q&A]
Any plans to add support for the IAB ad framework? We use display advertising on our website and the ad network provides their own CMP with all the IAB ad opt in/out choices. However, it's really ugly and means we need 2 popups on our site -- one for cookie consent, and one for ad choices. It would be awesome to integrate this all into one solution, especially since the cookieconset looks and operates so well.

Thanks for any ideas or opinions on this!

**Comment by orestbida** [comment]
Hi @rectifyer,

I'm not aware of the IAB ad framework. I don't know how it works or why it cannot be implemented using this plugin.

Can you provide a demo of how to configure IAB ads or — even better — its documentation?

**Comment by rectifyer** [comment]
I'm still learning about this myself, but here is what I'e learned so far. To be GDPR compliant with display ads, you need to have very detailed opt in/out for all potential ad networks you might use. This seems to get very complex, very quickly so just be aware of that 😄 

Here's a real world example for my website. When a a EU visitor views the site, they see an ad only CMP popup to choose their presences for how we can advertise to them. Once customized and approved, the CMP writes a `euconsent-v2` cookie with an encoded consent string. The ad network reads this 1st party cookie and adjusts the ads accordingly based on those preferences. In addition, we have our own cookie consent (using this project) to opt in/out of all the other non ad related cookies. I made this post since it would be awesome to unify all this into a single consent popup that provides detailed options for GDPR visitors.

This link was useful for understanding how other solutions handle it. It seems like there are several commercial consent companies that integrate the IAB ad stuff with their cookie consent into a unified popup.
https://www.iubenda.com/en/help/7440-iab-framework-cmp#what-is-the-iab-tcf

Here is the main project for the IAB framework.
https://github.com/InteractiveAdvertisingBureau/GDPR-Transparency-and-Consent-Framework

Here is a project about encoding/decoding the consent string.
https://github.com/InteractiveAdvertisingBureau/Consent-String-SDK-JS

This decoder is also useful to see what the cookie data is actually storing.
https://www.consentstringdecoder.com

Here are some screenshots of the provide CMP from our ad network. We can customize the CSS, but it's still not nearly as nice as the cookieconsent project. I believe all the vendors and settings can be dynamically loaded from some IAB JSON files, but it's still a huge amount of options for the end user.

Initial banner: https://capture.dropbox.com/9vfqs3S1Zc0mAWsj
Initial customize popup: https://capture.dropbox.com/H1hVYkMozp44UBzh
Vendor preferences: https://capture.dropbox.com/nXiXlzKXFyipbFBE
More settings: https://capture.dropbox.com/kUypT7qgwMDDfryq
More settings: https://capture.dropbox.com/T7MJrHXQzAJFCU4D

**Reply by orestbida** [reply]
@rectifyer thanks for the insight!

Unfortunately, the whole thing looks to be quite complex and costly

Based on this [implementation guideline](https://github.com/InteractiveAdvertisingBureau/GDPR-Transparency-and-Consent-Framework/blob/master/v1.1%20Implementation%20Guidelines.md#publisher-guidelines--) — and assuming that it is still valid — I'd have to register as a CMP solution, and the only way to do that is to [pay $1500/year](https://iabeurope.eu/join-the-tcf/).

Unless there is an official/formal way to integrate the IAB Consent Framework for free, I can safely say that it won't be implemented.

**Reply by rectifyer** [reply]
Oh wow, good catch on the $1500/year. That's unfortunate! If I find a way to integrate for free or get an exception for open source, I'll check back.
